dipy/denoise/localpca.py

```diff
@@ -0,0 +1,123 @@
+    theta = np.zeros(arr.shape, dtype=np.float64)
+    thetax = np.zeros(arr.shape, dtype=np.float64)
+
+    for k in range(patch_radius, arr.shape[2] - patch_radius, 1):
+        for j in range(patch_radius, arr.shape[1] - patch_radius, 1):
+            for i in range(patch_radius, arr.shape[0] - patch_radius, 1):
+
+                X = np.zeros(
+                    (patch_size * patch_size * patch_size, arr.shape[3]))
+                M = np.zeros(arr.shape[3])
+
+                temp = arr[i - patch_radius: i + patch_radius + 1,
+                           j - patch_radius: j + patch_radius + 1,
+                           k - patch_radius: k + patch_radius + 1, :]
+                X = temp.reshape(
+                    patch_size *
+                    patch_size *
+                    patch_size,
+                    arr.shape[3])
+                # compute the mean and normalize
+                M = np.mean(X, axis=0)
+                X = X - M
+                # PCA. There are two ways to do this:
+                # # 1. Compute the covariance matrix C = X_transpose X
+                # C = np.dot(X.T, X)
+                # C = C / X.shape[0]
+                # # and then decompose it:
+                # d, W = sla.eigh(C)
+
+                # 2. Alternatively, calculate the eigenvalues and
+                #  eigenvectors of the covariance matrix through an SVD:
+                # We can do this using scipy's linear algebra API (this is
+                # slower):
+                # U, S, Vt = sla.svd(X, full_matrices=False)
+                # Or through the bare bones direct access to LAPACK (faster):
+                U, S, Vt, info = sgesvd(X, 1, 0)
+                # These are the eigenvalues, but in ascending order:
+                d = S[::-1] ** 2 / X.shape[0]
+                # Rows are eigenvectors, but also in ascending eigenvalue
+                # order:
+                W = Vt[::-1].T
+
+                d[d < tau[i, j, k, :]] = 0
+                W_hat = np.zeros_like(W)
+                W_hat[:, d > 0] = W[:, d > 0]
+                Y = X.dot(W_hat)
+                X_est = Y.dot(np.transpose(W_hat))
+
+                # theta value
+                temp = X_est + \
+                    np.array([M, ] * X_est.shape[0], dtype=np.float64)
+                temp = temp.reshape(
+                    patch_size, patch_size, patch_size, arr.shape[3])
+                # Also update the estimate matrix which is X_est * theta
+
+                theta[i - patch_radius: i + patch_radius + 1,
+                      j - patch_radius: j + patch_radius + 1,
+                      k - patch_radius: k + patch_radius + 1,
+                      :] = theta[i - patch_radius: i + patch_radius + 1,
+                                 j - patch_radius: j + patch_radius + 1,
+                                 k - patch_radius: k + patch_radius + 1,
+                                 :] + 1.0 / (1.0 + np.linalg.norm(d,
```

Comment:
Are you absolutely sure you are looking at the most recent version of the code? This line, for example, was changed as you suggest in  https://github.com/nipy/dipy/pull/1223/commits/9c5de321b297803f2a8702cf91cd5df82454b6fa